File: scripts/update-university-data.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const DATA_DIR = path.join(ROOT, "data");
const CWUR_URL = "https://cwur.org/2026.php";
const CWUR_BASE_URL = "https://cwur.org/";

const SCHOOL_ALIASES = {
  "MIT": "Massachusetts Institute of Technology",
  "UC Berkeley": "University of California, Berkeley",
  "UCLA": "University of California, Los Angeles",
  "UC Irvine": "University of California, Irvine",
  "UC San Diego": "University of California, San Diego",
  "Georgia Tech": "Georgia Institute of Technology",
  "UT Austin": "University of Texas at Austin",
  "UIUC": "University of Illinois at Urbana-Champaign",
  "UW": "University of Washington"
};

const today = new Date().toISOString().slice(0, 10);

const programsPath = dataPath("programs.json");
const schoolsPath = dataPath("schools.json");
const windowsPath = dataPath("application-windows.json");
const rankingsPath = dataPath("school-rankings.json");
const reportPath = dataPath("update-report.json");

const programs = await readJson(programsPath);
const schools = await readJson(schoolsPath);
const applicationWindows = await readJson(windowsPath);
const currentRankings = await readJson(rankingsPath);

const validation = validateCatalog({ programs, schools, applicationWindows });
if (validation.errors.length > 0) {
  for (const error of validation.errors) {
    console.error(`[error] ${error}`);
  }
  process.exit(1);
}

const cwurRows = await fetchCwurRows(CWUR_URL);
const rankingResult = updateRankings(currentRankings, cwurRows, programs);
await writeJson(rankingsPath, rankingResult.rankingsPayload);

const report = {
  lastChecked: new Date().toISOString(),
  catalog: {
    programs: programs.length,
    schools: Object.keys(schools).length,
    applicationWindowEntries: Object.keys(applicationWindows).length,
    rankingEntries: Object.keys(rankingResult.rankingsPayload.rankings).length
  },
  cwur: {
    sourceUrl: CWUR_URL,
    rowsFetched: cwurRows.length,
    matchedSchools: rankingResult.matchedSchools.length,
    unmatchedProgramSchools: rankingResult.unmatchedProgramSchools,
    updatedRankingKeys: rankingResult.updatedRankingKeys
  },
  validation
};

await writeJson(reportPath, report);

console.log(`Catalog validated: ${programs.length} programs, ${Object.keys(schools).length} schools.`);
console.log(`CWUR rows fetched: ${cwurRows.length}; matched program schools: ${rankingResult.matchedSchools.length}.`);
if (rankingResult.unmatchedProgramSchools.length > 0) {
  console.log(`Unmatched program schools: ${rankingResult.unmatchedProgramSchools.join(", ")}`);
}

function dataPath(filename) {
  return path.join(DATA_DIR, filename);
}

async function readJson(filename) {
  const text = await fs.readFile(filename, "utf8");
  return JSON.parse(text);
}

async function writeJson(filename, value) {
  await fs.writeFile(filename, `${JSON.stringify(value, null, 2)}\n`, "utf8");
}

function validateCatalog({ programs, schools, applicationWindows }) {
  const errors = [];
  const warnings = [];
  const ids = new Set();
  const schoolNames = new Set(Object.keys(schools));
  const usedSchools = new Set();

  for (const program of programs) {
    const label = program?.id ?? "(missing id)";

    if (!program?.id || typeof program.id !== "string") {
      errors.push("Program is missing a string id.");
      continue;
    }

    if (ids.has(program.id)) {
      errors.push(`Duplicate program id: ${program.id}`);
    }
    ids.add(program.id);

    if (!program.school || !schoolNames.has(program.school)) {
      errors.push(`${label} references missing school profile: ${program.school}`);
    } else {
      usedSchools.add(program.school);
    }

    if (!applicationWindows[program.id]) {
      errors.push(`${label} is missing application windows.`);
    }

    for (const key of ["program", "admission", "international", "apply"]) {
      const url = program.links?.[key];
      if (!isHttpUrl(url)) {
        errors.push(`${label} has invalid ${key} link.`);
      }
    }
  }

  for (const [programId, windows] of Object.entries(applicationWindows)) {
    if (!ids.has(programId)) {
      warnings.push(`Application windows exist for unknown program id: ${programId}`);
    }

    if (!Array.isArray(windows) || windows.length === 0) {
      errors.push(`${programId} must have at least one application window.`);
      continue;
    }

    for (const window of windows) {
      if (!window.intake || typeof window.intake !== "string") {
        errors.push(`${programId} has a window without intake.`);
      }

      if (window.opens !== null && !isMonthDay(window.opens)) {
        errors.push(`${programId} has invalid opens value: ${window.opens}`);
      }

      if (!isMonthDay(window.deadline)) {
        errors.push(`${programId} has invalid deadline value: ${window.deadline}`);
      }
    }
  }

  for (const [school, profile] of Object.entries(schools)) {
    if (!usedSchools.has(school)) {
      warnings.push(`School profile is not used by any program: ${school}`);
    }

    if (!["Public", "Private"].includes(profile.schoolType)) {
      errors.push(`${school} has invalid schoolType: ${profile.schoolType}`);
    }

    if (!isHttpUrl(profile.sourceUrl)) {
      errors.push(`${school} has invalid sourceUrl.`);
    }
  }

  return { errors, warnings };
}

function isHttpUrl(value) {
  return typeof value === "string" && /^https?:\/\/\S+$/i.test(value.trim());
}

function isMonthDay(value) {
  if (typeof value !== "string" || !/^\d{2}-\d{2}$/.test(value)) {
    return false;
  }

  const [month, day] = value.split("-").map(Number);
  if (month < 1 || month > 12 || day < 1 || day > 31) {
    return false;
  }

  const date = new Date(Date.UTC(2024, month - 1, day));
  return date.getUTCMonth() === month - 1 && date.getUTCDate() === day;
}

async function fetchCwurRows(url) {
  const response = await fetch(url, {
    headers: {
      "User-Agent": "AdmitFlow data updater (+https://admit-flow.com)",
      "Accept": "text/html,application/xhtml+xml"
    }
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch CWUR rankings: ${response.status} ${response.statusText}`);
  }

  const html = await response.text();
  const rows = [];
  const rowPattern = /<tr>([\s\S]*?)<\/tr>/gi;
  let rowMatch;

  while ((rowMatch = rowPattern.exec(html))) {
    const cells = Array.from(rowMatch[1].matchAll(/<td[^>]*>([\s\S]*?)<\/td>/gi), (match) => match[1]);
    if (cells.length < 9) {
      continue;
    }

    const rank = firstNumber(stripHtml(cells[0]));
    const institutionLink = cells[1].match(/<a\s+href="([^"]+)"[^>]*>([\s\S]*?)<\/a>/i);
    const location = stripHtml(cells[2]);
    const nationalRank = firstNumber(stripHtml(cells[3]));
    const score = Number(stripHtml(cells[8]));

    if (!rank || !institutionLink || location !== "USA" || !nationalRank || Number.isNaN(score)) {
      continue;
    }

    const href = decodeHtml(institutionLink[1]);
    const institution = stripHtml(institutionLink[2]);

    rows.push({
      institution,
      normalizedInstitution: normalizeName(institution),
      worldRank: rank,
      nationalRank,
      score,
      sourceUrl: new URL(href, CWUR_BASE_URL).href
    });
  }

  if (rows.length === 0) {
    throw new Error("CWUR parser did not find any USA ranking rows.");
  }

  return rows;
}

function updateRankings(currentRankings, cwurRows, programs) {
  const rowsByName = new Map(cwurRows.map((row) => [row.normalizedInstitution, row]));
  const programSchools = Array.from(new Set(programs.map((program) => program.school))).sort();
  const existingEntries = currentRankings.rankings ?? {};
  const updatedRankings = {};
  const matchedSchools = [];
  const unmatchedProgramSchools = [];
  const updatedRankingKeys = [];

  for (const school of programSchools) {
    const existing = existingEntries[school];
    const candidates = [
      existing?.sourceInstitution,
      SCHOOL_ALIASES[school],
      school
    ].filter(Boolean);

    const row = candidates
      .map((candidate) => rowsByName.get(normalizeName(candidate)))
      .find(Boolean);

    if (!row) {
      unmatchedProgramSchools.push(school);
      if (existing) {
        updatedRankings[school] = existing;
      }
      continue;
    }

    const nextEntry = {
      sourceInstitution: row.institution,
      worldRank: row.worldRank,
      nationalRank: row.nationalRank,
      score: row.score,
      sourceUrl: row.sourceUrl
    };

    updatedRankings[school] = nextEntry;
    matchedSchools.push(school);

    if (JSON.stringify(existing) !== JSON.stringify(nextEntry)) {
      updatedRankingKeys.push(school);
    }
  }

  for (const [school, entry] of Object.entries(existingEntries)) {
    if (!updatedRankings[school]) {
      updatedRankings[school] = entry;
    }
  }

  const sortedRankings = Object.fromEntries(
    Object.entries(updatedRankings).sort(([, left], [, right]) => {
      return left.worldRank - right.worldRank || left.sourceInstitution.localeCompare(right.sourceInstitution);
    })
  );

  return {
    rankingsPayload: {
      source: {
        ...(currentRankings.source ?? {}),
        name: "Center for World University Rankings (CWUR)",
        edition: "2026 Global 2000",
        url: CWUR_URL,
        country: "USA",
        updated: today
      },
      rankings: sortedRankings
    },
    matchedSchools,
    unmatchedProgramSchools,
    updatedRankingKeys
  };
}

function firstNumber(value) {
  const match = String(value).match(/\d+/);
  return match ? Number(match[0]) : null;
}

function stripHtml(value) {
  return decodeHtml(String(value).replace(/<[^>]*>/g, "")).replace(/\s+/g, " ").trim();
}

function decodeHtml(value) {
  return String(value)
    .replace(/&nbsp;/g, " ")
    .replace(/&amp;/g, "&")
    .replace(/&#039;/g, "'")
    .replace(/&quot;/g, "\"")
    .replace(/&ndash;|&mdash;/g, "-");
}

function normalizeName(value) {
  return String(value)
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/&/g, " and ")
    .replace(/[\u2010-\u2015]/g, "-")
    .replace(/[^a-zA-Z0-9]+/g, " ")
    .replace(/\bthe\b/g, " ")
    .replace(/\s+/g, " ")
    .trim()
    .toLowerCase();
}

File: scripts/update_university_data.py
from __future__ import annotations

import re
from pathlib import Path
from typing import Any
from urllib.parse import urljoin

import requests
from bs4 import BeautifulSoup

from utils import DATA_DIR, is_http_url, normalize_name, read_json, today_iso, utc_now, write_json

CWUR_URL = "https://cwur.org/2026.php"
CWUR_BASE_URL = "https://cwur.org/"

SCHOOL_ALIASES = {
    "MIT": "Massachusetts Institute of Technology",
    "UC Berkeley": "University of California, Berkeley",
    "UCLA": "University of California, Los Angeles",
    "UC Irvine": "University of California, Irvine",
    "UC San Diego": "University of California, San Diego",
    "Georgia Tech": "Georgia Institute of Technology",
    "UT Austin": "University of Texas at Austin",
    "UIUC": "University of Illinois at Urbana-Champaign",
    "UW": "University of Washington",
}


def main() -> None:
    programs = read_json(DATA_DIR / "programs.json")
    schools = read_json(DATA_DIR / "schools.json")
    application_windows = read_json(DATA_DIR / "application-windows.json")
    current_rankings = read_json(DATA_DIR / "school-rankings.json")

    validation = validate_catalog(programs, schools, application_windows)
    if validation["errors"]:
        for error in validation["errors"]:
            print(f"[error] {error}")
        raise SystemExit(1)

    cwur_rows = fetch_cwur_rows(CWUR_URL)
    ranking_result = update_rankings(current_rankings, cwur_rows, programs)
    write_json(DATA_DIR / "school-rankings.json", ranking_result["rankings_payload"])

    report = {
        "lastChecked": utc_now(),
        "catalog": {
            "programs": len(programs),
            "schools": len(schools),
            "applicationWindowEntries": len(application_windows),
            "rankingEntries": len(ranking_result["rankings_payload"]["rankings"]),
        },
        "cwur": {
            "sourceUrl": CWUR_URL,
            "rowsFetched": len(cwur_rows),
            "matchedSchools": len(ranking_result["matched_schools"]),
            "unmatchedProgramSchools": ranking_result["unmatched_program_schools"],
            "updatedRankingKeys": ranking_result["updated_ranking_keys"],
        },
        "validation": validation,
    }
    write_json(DATA_DIR / "update-report.json", report)

    print(f"Catalog validated: {len(programs)} programs, {len(schools)} schools.")
    print(
        f"CWUR rows fetched: {len(cwur_rows)}; "
        f"matched program schools: {len(ranking_result['matched_schools'])}."
    )
    if ranking_result["unmatched_program_schools"]:
        print(f"Unmatched program schools: {', '.join(ranking_result['unmatched_program_schools'])}")


def validate_catalog(
    programs: list[dict[str, Any]],
    schools: dict[str, dict[str, Any]],
    application_windows: dict[str, list[dict[str, Any]]],
) -> dict[str, list[str]]:
    errors: list[str] = []
    warnings: list[str] = []
    ids: set[str] = set()
    school_names = set(schools.keys())
    used_schools: set[str] = set()

    for program in programs:
        program_id = program.get("id")
        label = program_id or "(missing id)"

        if not isinstance(program_id, str) or not program_id:
            errors.append("Program is missing a string id.")
            continue

        if program_id in ids:
            errors.append(f"Duplicate program id: {program_id}")
        ids.add(program_id)

        school = program.get("school")
        if not school or school not in school_names:
            errors.append(f"{label} references missing school profile: {school}")
        else:
            used_schools.add(school)

        if program_id not in application_windows:
            errors.append(f"{label} is missing application windows.")

        links = program.get("links") or {}
        for key in ("program", "admission", "international", "apply"):
            if not is_http_url(links.get(key)):
                errors.append(f"{label} has invalid {key} link.")

    for program_id, windows in application_windows.items():
        if program_id not in ids:
            warnings.append(f"Application windows exist for unknown program id: {program_id}")

        if not isinstance(windows, list) or not windows:
            errors.append(f"{program_id} must have at least one application window.")
            continue

        for window in windows:
            if not isinstance(window.get("intake"), str) or not window["intake"]:
                errors.append(f"{program_id} has a window without intake.")
            if window.get("opens") is not None and not is_month_day(window.get("opens")):
                errors.append(f"{program_id} has invalid opens value: {window.get('opens')}")
            if not is_month_day(window.get("deadline")):
                errors.append(f"{program_id} has invalid deadline value: {window.get('deadline')}")

    for school, profile in schools.items():
        if school not in used_schools:
            warnings.append(f"School profile is not used by any program: {school}")
        if profile.get("schoolType") not in {"Public", "Private"}:
            errors.append(f"{school} has invalid schoolType: {profile.get('schoolType')}")
        if not is_http_url(profile.get("sourceUrl")):
            errors.append(f"{school} has invalid sourceUrl.")

    return {"errors": errors, "warnings": warnings}


def is_month_day(value: Any) -> bool:
    if not isinstance(value, str) or re.match(r"^\d{2}-\d{2}$", value) is None:
        return False

    month, day = [int(part) for part in value.split("-")]
    if month < 1 or month > 12 or day < 1 or day > 31:
        return False

    month_lengths = {1: 31, 2: 29, 3: 31, 4: 30, 5: 31, 6: 30, 7: 31, 8: 31, 9: 30, 10: 31, 11: 30, 12: 31}
    return day <= month_lengths[month]


def fetch_cwur_rows(url: str) -> list[dict[str, Any]]:
    response = requests.get(
        url,
        headers={
            "User-Agent": "AdmitFlow data updater (+https://admit-flow.com)",
            "Accept": "text/html,application/xhtml+xml",
        },
        timeout=20,
    )
    response.raise_for_status()

    soup = BeautifulSoup(response.text, "html.parser")
    rows: list[dict[str, Any]] = []

    for table_row in soup.select("tr"):
        cells = table_row.find_all("td")
        if len(cells) < 9:
            continue

        institution_link = cells[1].find("a")
        if not institution_link:
            continue

        world_rank = first_number(cells[0].get_text(" ", strip=True))
        institution = institution_link.get_text(" ", strip=True)
        location = cells[2].get_text(" ", strip=True)
        national_rank = first_number(cells[3].get_text(" ", strip=True))
        score_text = cells[8].get_text(" ", strip=True)

        try:
            score = float(score_text)
        except ValueError:
            continue

        if not world_rank or location != "USA" or not national_rank:
            continue

        rows.append(
            {
                "institution": institution,
                "normalizedInstitution": normalize_name(institution),
                "worldRank": world_rank,
                "nationalRank": national_rank,
                "score": score,
                "sourceUrl": urljoin(CWUR_BASE_URL, institution_link.get("href", "")),
            }
        )

    if not rows:
        raise RuntimeError("CWUR parser did not find any USA ranking rows.")

    return rows


def update_rankings(
    current_rankings: dict[str, Any],
    cwur_rows: list[dict[str, Any]],
    programs: list[dict[str, Any]],
) -> dict[str, Any]:
    rows_by_name = {row["normalizedInstitution"]: row for row in cwur_rows}
    program_schools = sorted({program["school"] for program in programs})
    existing_entries = current_rankings.get("rankings") or {}
    updated_rankings: dict[str, dict[str, Any]] = {}
    matched_schools: list[str] = []
    unmatched_program_schools: list[str] = []
    updated_ranking_keys: list[str] = []

    for school in program_schools:
        existing = existing_entries.get(school)
        candidates = [
            existing.get("sourceInstitution") if isinstance(existing, dict) else None,
            SCHOOL_ALIASES.get(school),
            school,
        ]
        row = next((rows_by_name.get(normalize_name(candidate)) for candidate in candidates if candidate), None)

        if not row:
            unmatched_program_schools.append(school)
            if existing:
                updated_rankings[school] = existing
            continue

        next_entry = {
            "sourceInstitution": row["institution"],
            "worldRank": row["worldRank"],
            "nationalRank": row["nationalRank"],
            "score": row["score"],
            "sourceUrl": row["sourceUrl"],
        }
        updated_rankings[school] = next_entry
        matched_schools.append(school)

        if existing != next_entry:
            updated_ranking_keys.append(school)

    for school, entry in existing_entries.items():
        if school not in updated_rankings:
            updated_rankings[school] = entry

    sorted_rankings = dict(
        sorted(
            updated_rankings.items(),
            key=lambda item: (
                item[1].get("worldRank", 999999),
                item[1].get("sourceInstitution", ""),
            ),
        )
    )

    return {
        "rankings_payload": {
            "source": {
                **(current_rankings.get("source") or {}),
                "name": "Center for World University Rankings (CWUR)",
                "edition": "2026 Global 2000",
                "url": CWUR_URL,
                "country": "USA",
                "updated": today_iso(),
            },
            "rankings": sorted_rankings,
        },
        "matched_schools": matched_schools,
        "unmatched_program_schools": unmatched_program_schools,
        "updated_ranking_keys": updated_ranking_keys,
    }


def first_number(value: str) -> int | None:
    match = re.search(r"\d+", value)
    return int(match.group(0)) if match else None


if __name__ == "__main__":
    main()
